### PR TITLE
CI: Enable Qt5 + lightpreview

### PR DESCRIPTION
**Pull request by jonathanlinat**
This PR primarily focuses on facilitating the building and releasing of `lightpreview` for all operating systems using Qt5.

Additionally, it enhances the runners, updates the external actions, and improves the clarity of the existing steps.

### CODE REVIEW

**Comment by jonathanlinat** [comment]
@ericwa It seems you need to approve this PR to let GitHub Actions execute the corresponding workflow.

File: .github/workflows/cmake.yml [hunk 1]
```diff
@@ -1,86 +0,0 @@
...
-    - name: Setup MSVC environment
-      if: runner.os == 'Windows' 
-      uses: ilammy/msvc-dev-cmd@v1
-
-    - name: Linux Build
-      if: runner.os == 'Linux'
-      run: |
-          sudo apt-get update
-          # for Sphinx
-          sudo apt-get install -y python3-pip
```
Comment (jonathanlinat) [inline .github/workflows/cmake.yml:35]: I removed this as it is already part of the runner dependencies.

File: .github/workflows/cmake.yml [hunk 2]
```diff
@@ -1,86 +0,0 @@
...
-on: [push, pull_request]
-
-jobs:
-  build:
-    runs-on: ${{ matrix.os }}
-    strategy:
-      # Don't cancel the macOS build if the Linux build fails, etc.
-      fail-fast: false
-      matrix:
-        os: [ubuntu-20.04, macos-11, windows-2019]
```
Comment (jonathanlinat) [inline .github/workflows/cmake.yml:12]: I upgraded them to version `22.04`, `12` and `2022` respectively.

File: .github/workflows/cmake.yml [hunk 3]
```diff
@@ -1,86 +0,0 @@
-name: CMake
-
-on: [push, pull_request]
-
-jobs:
-  build:
-    runs-on: ${{ matrix.os }}
-    strategy:
-      # Don't cancel the macOS build if the Linux build fails, etc.
```
Comment (jonathanlinat) [inline .github/workflows/cmake.yml:9]: I removed all the comments as the steps must be clear enough to understand what's happening by themselves.

File: .github/workflows/cmake.yml [hunk 4]
```diff
@@ -1,86 +0,0 @@
-name: CMake
```
Comment (jonathanlinat) [inline .github/workflows/cmake.yml:1]: I renamed the workflow to `Continuous Building`.

File: .github/workflows/cmake.yml [hunk 5]
```diff
@@ -1,86 +0,0 @@
...
-        os: [ubuntu-20.04, macos-11, windows-2019]
-        use-asan: [YES, NO]
-        exclude:
-          - os: windows-2019
-            use-asan: YES
-    env:
-        # Expose to the build-*.sh in an environment variable
-        USE_ASAN: ${{ matrix.use-asan }}
-    steps:
-    - uses: actions/checkout@v2
```
Comment (jonathanlinat) [inline .github/workflows/cmake.yml:21]: I updated this action to its latest available version (`v4`).

File: .github/workflows/cmake.yml [hunk 6]
```diff
@@ -1,86 +0,0 @@
...
-      with:
-        submodules: 'recursive'
-
-    # https://github.com/ilammy/msvc-dev-cmd
-    - name: Setup MSVC environment
-      if: runner.os == 'Windows'
```
Comment (jonathanlinat) [inline .github/workflows/cmake.yml:27]: I use this instead `startsWith(matrix.os, 'windows-')`.

File: .github/workflows/cmake.yml [hunk 7]
```diff
@@ -1,86 +0,0 @@
...
-      run: |
-          .\build-windows.ps1
-
-    # Upload artifacts.
-    # These need to be separate, otherwise all of the artifacts are bundled into
-    # one .zip file.
-
-    - name: Upload win64 artifact
-      uses: actions/upload-artifact@v3
```
Comment (jonathanlinat) [inline .github/workflows/cmake.yml:53]: I updated this action to its latest available version (`v4`).

File: .github/workflows/cmake.yml [hunk 8]
```diff
@@ -1,86 +0,0 @@
...
-    - uses: actions/checkout@v2
-      with:
-        submodules: 'recursive'
-
-    # https://github.com/ilammy/msvc-dev-cmd
-    - name: Setup MSVC environment
-      if: runner.os == 'Windows' 
-      uses: ilammy/msvc-dev-cmd@v1
-
-    - name: Linux Build
```
Comment (jonathanlinat) [inline .github/workflows/cmake.yml:30]: I introduced the following naming convention for better understanding: `[Operating System]: [Concise title of the step]`

File: .github/workflows/cmake.yml [hunk 9]
```diff
@@ -1,86 +0,0 @@
...
-          ./build-osx.sh
-
-    - name: Windows Build
-      if: runner.os == 'Windows'
-      run: |
-          .\build-windows.ps1
-
-    # Upload artifacts.
-    # These need to be separate, otherwise all of the artifacts are bundled into
-    # one .zip file.
```
Comment (jonathanlinat) [inline .github/workflows/cmake.yml:50]: Removed.

File: .github/workflows/cmake.yml [hunk 10]
```diff
@@ -1,86 +0,0 @@
...
-      uses: actions/upload-artifact@v3
-      if: ${{ runner.os == 'macOS' && matrix.use-asan == 'NO' }}
-      with:
-        name: ericw-tools-${{ github.sha }}-macOS
-        path: |
-          build-osx/ericw-tools-*.zip
-
-    - name: Create GitHub Release and upload builds
-      uses: softprops/action-gh-release@v1
-      if: ${{ startsWith(github.ref, 'refs/tags/') && matrix.use-asan == 'NO' && runner.os != 'Windows' }}
```
Comment (jonathanlinat) [inline .github/workflows/cmake.yml:78]: I removed this part `&& runner.os != 'Windows'`.

File: .github/workflows/continuous-building.yml [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+    steps:
+    - uses: actions/checkout@v4
+      with:
+        submodules: recursive
+
+    - name: 'Linux: Install Qt5'
+      if: startsWith(matrix.os, 'ubuntu-')
+      run: |
+        sudo apt update
+        sudo apt install qtbase5-dev libqt5svg5-dev
```
Comment (jonathanlinat) [inline .github/workflows/continuous-building.yml:35]: Linux: This is where the magic happens.

File: .github/workflows/continuous-building.yml [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+      uses: actions/upload-artifact@v4
+      with:
+        name: ericw-tools-${{ github.sha }}-linux
+        path: build-linux/*-Linux.zip
+
+    - name: 'macOS: Install Qt5'
+      if: ${{ startsWith(matrix.os, 'macos-') }}
+      uses: jurplel/install-qt-action@v3
+      with:
+        version: 5.15.2
```
Comment (jonathanlinat) [inline .github/workflows/continuous-building.yml:50]: macOs: This is where the magic happens.

File: .github/workflows/continuous-building.yml [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+
+    - name: 'Windows: Setup MSVC environment'
+      if: startsWith(matrix.os, 'windows-')
+      uses: ilammy/msvc-dev-cmd@v1
+    - name: 'Windows: Install Qt5'
+      if: ${{ startsWith(matrix.os, 'windows-') }}
+      uses: jurplel/install-qt-action@v3
+      with:
+        version: 5.15.2
+        arch: win64_msvc2019_64
```
Comment (jonathanlinat) [inline .github/workflows/continuous-building.yml:69]: Windows: This is where the magic happens.

**Comment by jonathanlinat** [comment]
@ericwa i'll take a look tonight and get back to you.

**Comment by ericwa** [comment]
I downloaded the windows build and tested lightpreview and it seems to work fine, so that's a good start! We can drop the use of appveyor for releases now. We could remove appveyor entirely - it takes about 2x as long to build as GH actions does. But GH actions is less convenient for people trying out dev builds - requires login, the .zips are wrapped in multiple layers of zip and extra folders, etc. But I also don't like maintaining two different flavors of Windows build (GH actions / appveyor) and splitting testing between them - it means we might miss an issue in a Windows GH release that didn't show up in the Appveyor dev builds for some reason (e.g. different compiler version), so that's an argument for dropping appveyor entirely.

I'll test the Linux artifact in a VM in a bit.

For the mac one, I don't think `lightpreview` will work because we're not copying the Qt libraries into the archive. We don't need to fix it right now.. I think it's going to be hard to get lightpreview packaging unless I get a new mac.

**Comment by ericwa** [comment]
Seems like the linux build is failing because CPack is packaging the archive as `merge-Linux.zip`

```
CPack: Create package using ZIP
CPack: Install projects
CPack: - Run preinstall target for: ericw-tools
CPack: - Install project: ericw-tools []
CPack: Create package
CPack: - package: /home/runner/work/ericw-tools/ericw-tools/build-linux/merge-Linux.zip 
```

and then later in the `build-linux-64.sh` we try to unzip it as: `unzip -X ericw-tools-*.zip`

**Comment by ericwa** [comment]
The linux build is looking good now. I'm testing it on Ubuntu 23.04 since that's the VM I had available, and the `lightpreview` binary works (using the system copy of Qt 5).

**Comment by jonathanlinat** [comment]
@ericwa That's great and I am glad it works. 👌🏻

I am pretty sure we can replicate the same for macOS.

**Comment by jonathanlinat** [comment]
> The linux build is looking good now. I'm testing it on Ubuntu 23.04 since that's the VM I had available, and the `lightpreview` binary works (using the system copy of Qt 5).

I confirm it works (22.04).

image

**Comment by jonathanlinat** [comment]
> I downloaded the windows build and tested lightpreview and it seems to work fine, so that's a good start! We can drop the use of appveyor for releases now. We could remove appveyor entirely - it takes about 2x as long to build as GH actions does. But GH actions is less convenient for people trying out dev builds - requires login, the .zips are wrapped in multiple layers of zip and extra folders, etc. But I also don't like maintaining two different flavors of Windows build (GH actions / appveyor) and splitting testing between them - it means we might miss an issue in a Windows GH release that didn't show up in the Appveyor dev builds for some reason (e.g. different compiler version), so that's an argument for dropping appveyor entirely.
> 
> I'll test the Linux artifact in a VM in a bit.
> 
> For the mac one, I don't think `lightpreview` will work because we're not copying the Qt libraries into the archive. We don't need to fix it right now.. I think it's going to be hard to get lightpreview packaging unless I get a new mac.

Regarding the nested ZIP files, I can take a look at it. I've made the corresponding changes to my starter kit to ensure the `doc` and `bin` directories are the direct content of the downloadable assets. It's not a difficult task to achieve.

**Comment by ericwa** [comment]
> > [...] the .zips are wrapped in multiple layers of zip and extra folders [...]
> 
> Regarding the nested ZIP files, I can take a look at them, if that's okay with you. I've made the corresponding changes to my starter kit to ensure the `doc` and `bin` directories are the direct content of the downloadable assets. It's a relatively easy task to achieve.

Sure, if it's easy, thanks!

Finally got the CI to produce a working lightpreview.app (this is on 10.15, x86_64):
Screen Shot 2024-01-08 at 12 54 06 AM
Also fixed the command-line tools which had been broken on macOS since 2.0.0-alpha1 due to an missing rpath setting.

I'll confirm tomorrow that I didn't break the Linux/Windows builds.
Other than that I think this is good to merge whenever you're ready.

**Comment by jonathanlinat** [comment]
@ericwa I introduced the additional steps to re-pack the binaries as proposed.

Additionally, I decided to remove the last step from this PR. I'll create another PR with a new workflow to include a release job (`continuous-releasing`) for better organization.

**Comment by jonathanlinat** [comment]
Here we go.

image

I would also recommend adding, in another PR, an extra step for each system to include the README and LICENSE files from the repository in the artifacts (ZIP file), together with the `doc` and `bin` directories.

**Comment by ericwa** [comment]
all good if I squash and merge @jonathanlinat ?

**Comment by jonathanlinat** [comment]
> all good if I squash and merge @jonathanlinat ?

Absolutely.

**Comment by ericwa** [comment]
thanks for this!

**Pull request merged**